src/react-render-perf/lesson-03/README.md

```diff
@@ -2,4 +2,75 @@
 
 # 03 - Avoid Using Context
 
-TODO
+If your context contains an event emitter (or other class instance) and it's only 
+used once in your application, you can simplify things even further by using a
+singleton instance of it.
```

Comment:
ahhhhh I ... don't like this? I would be very surprised if it resulted in meaningful performance improvements.
While it does make testing much more complicated, the bigger problem in my mind is the proliferation of "spooky action at a distance"; a global singleton is accessible by any code in any context, whereas a `Context` is only accessible in React code under `hooks`. I'd rather like to discourage making singletons in general, and I think the performance case for them here isn't terribly robust.

Reply:
That's a fair point.  @somewhatabstract what are your thoughts on this?

Reply:
Another nice thing about it being in a React Context is that it's easy to mock in test.

Reply:
I have found and observed that using context correctly to avoid unnecessary rendering has proven to be hard, especially when the data being provided is mutable. Much of the work in Khanmigo was down to bad context implementation. Writing tests to verify that context is being properly handled is not terribly hard but not a common pattern that folks are familiar with - I have done it and still have to look it up, some folks just don't do it.

In addition, context is often ONLY available in React as you point out, @jaredly, which often becomes a sticking point because there can be situations where we need to access that data outside of React and therefore, have to do a reasonable big refactor that could be avoided by just not using it.

Finally, with testing, context becomes cumbersome. For things consuming the context, it requires nested providers or a test harness adapter to mitigate that, and often some complex mocking to have things work in a testable way that is just much easier when it's a plain old JavaScript thing. For the context itself, if it is more than just static data, it adds extra layers of abstraction that complicate testing (and often end up being made easier by putting the context implementation into a class or set of functions that can be tested outside the React world - which is a short step from a singleton).

In the end, for very simple scenarios, I think React context can make sense, but I think it's a pattern that all too easily is a foot gun; much more so than singletons or other methods of managing this kind of shared state.

Overall, I think the best thing is to avoid both - but there are circumstances where that isn't possible and the more complex the context, I have found the more cumbersome and foot-gunny it becomes.

Reply:
@somewhatabstract thanks for the additional context. 😉 I'm going to update the README.md to talk about some of the pros/cons of using a singleton vs context.

Reply:
that's compelling! I withdraw my objection.
also potentially interesting to look into, I've heard really good things about [jotai](https://jotai.org/)

Reply:
> also potentially interesting to look into, I've heard really good things about [jotai](https://jotai.org/)

Someone mentioned this recently at ACM. It does look worth a deeper investigation. Perhaps one or more of us can time box an investigation and do an eng lightning talk on it?

Reply:
I think it would be good to have a look at zustand, recoil.js, and mobx and see how they stack up to jotai.